### PR TITLE
Rename kernel reserved function names

### PR DESCRIPTION
**Pull request by pfl**
This patch uses the scripts created in #12086 to rename reserved function names in the kernel/ directory and applying the changes found to the rest of the source code tree. The prefixes that are left as-is are '_Pragma', '_pragma', '__atomic', '__builtin' and '__SYSCALL', where the last one is left alone only due to practical reasons, there is already a macro '_SYSCALL' that gets renamed to 'Z_SYSCALL'.

The .c and .h files are updated, and the code compiles, but some of the linking is not yet detected, so this isn't fully working yet. An attempt to hack up on the syscall generating script has also been done.

This patch set is part of #9882.

### CODE REVIEW

**Review by andyross — DISMISSED** [review]
No complaints in general.   One thing that needs to be fixed is that (if I'm understanding it right) the series is not bisectable.  The build breaks between the second patch (which renames the APIs) and the third (which changes usage).  We really need those to be simultaneous.  If the change becomes too large for a single patch, you can break it up by API instead of directory.

A few notes on the naming chosen:

* Some of our more oddball naming should be fixed manually, I think.  "z_Cstart()" and "Swap()" aren't very good names (and weren't before).  I'd suggest z_swap() and z_cstart().

* Naming that has existing prefixes should probably combine that with the "z-ification" instead of just making the prefix longer.  The "_impl_" prefix is already applied to a symbol name (e.g. k_frog_dissect()) that is namespaced, so I'd argue we don't want to spend the bytes on a separate "z_" and should just call it "impl_*()".

* Similarly "z_arch_" could just be "zarch_" and save a byte.

**Comment by pfl** [comment]
Yes, these should be squashed together, bisectability is quite dead in between. It is in two parts now so that one has some chance to figure out which output comes from which step, should someone be interested in dissecting brokenness. The naming can be changed, the scripts work decently well so they can be updated and the procedure re-applied. Re-running the scripts needs to happen already now, there have been some changes in the source tree already causing conflicts.

The naming and prefixes could be discussed for example in #9882.

File: arch/arc/core/prep_c.c [hunk 1]
```diff
@@ -119,8 +119,8 @@ void _PrepC(void)
 {
 	_icache_setup();
 	adjust_vector_table_base();
-	_bss_zero();
-	_data_copy();
-	_Cstart();
+	z_bss_zero();
+	data_copy();
```
Comment (andrewboie) [inline arch/arc/core/prep_c.c]: z_data_copy()?

File: arch/arc/core/thread.c [hunk 2]
```diff
@@ -131,7 +131,7 @@ void _new_thread(struct k_thread *thread, k_thread_stack_t *stack,
 #endif
 	}
 
-	_new_thread_init(thread, pStackMem, stackSize, priority, options);
+	new_thread_init(thread, pStackMem, stackSize, priority, options);
```
Comment (andrewboie) [inline arch/arc/core/thread.c]: z_new_thread_init

File: arch/arc/include/kernel_arch_func.h [hunk 3]
```diff
@@ -37,7 +37,7 @@ static ALWAYS_INLINE void kernel_arch_init(void)
 }
 
 static ALWAYS_INLINE void
-_set_thread_return_value(struct k_thread *thread, unsigned int value)
+set_thread_return_value(struct k_thread *thread, unsigned int value)
```
Comment (andrewboie) [inline arch/arc/include/kernel_arch_func.h]: z_set_thread_return_value

**Review by tyrantlizard77 — CHANGES_REQUESTED** [review]
There's something messed up in the script which is removing leading underscores instead of changing them to z_

File: scripts/gen_syscall_header.py [hunk 4]
```diff
@@ -88,7 +88,7 @@ def gen_make_syscall(ret, argc, tabcount):
 def gen_call_impl(ret, argc):
     if (ret != Retval.VOID):
         sys.stdout.write("return ")
-    sys.stdout.write("_impl_##name(")
+    sys.stdout.write("z_sys_impl_##name(")
```
Comment (andrewboie) [inline scripts/gen_syscall_header.py]: why not z_impl_?

I don't like the additional _sys_ here, the names are long enough already

File: arch/arc/core/prep_c.c (hunk 1 above)
Comment (pfl) [reply arch/arc/core/prep_c.c]: data_copy() comes from the previous kernel patch in file kernel/include/kernel_internal.h:
```
 #ifdef CONFIG_XIP
-void _data_copy(void);
+void data_copy(void);
 #else
-static inline void _data_copy(void)
+static inline void data_copy(void)
```
Coccinelle sees the static inline void _data_copy() and correctly renames it data_copy(). Of course the above wants it the opposite way.

File: arch/arc/core/thread.c (hunk 2 above)
Comment (pfl) [reply arch/arc/core/thread.c]: in kernel/device.c we have:
```
-static ALWAYS_INLINE void _new_thread_init(struct k_thread *thread,
+static ALWAYS_INLINE void new_thread_init(struct k_thread *thread,
```

File: arch/arc/include/kernel_arch_func.h (hunk 3 above)
Comment (pfl) [reply arch/arc/include/kernel_arch_func.h]: As above.

File: scripts/gen_syscall_header.py (hunk 4 above)
Comment (pfl) [reply scripts/gen_syscall_header.py]: z_sys_impl_ was taken from #9882, we can change it so something more reasonable.

**Comment by pfl** [comment]
> There's something messed up in the script which is removing leading underscores instead of changing them to z_

Yes. It seems the code makes functions static when a certain #define is in place and global when the #define is not there. I suppose we do want to have all functions global and let the linker/compiler remove them when empty?

File: arch/arc/core/prep_c.c [hunk 5]
```diff
@@ -105,7 +105,7 @@ static void adjust_vector_table_base(void)
 	}
 }
 
-extern FUNC_NORETURN void _Cstart(void);
+extern FUNC_NORETURN void z_Cstart(void);
```
Comment (ceolin) [inline arch/arc/core/prep_c.c]: I second Andy's comment to use z_cstart() and not using camelcase.

File: arch/arc/core/thread.c (hunk 2 above)
Comment (ceolin) [reply arch/arc/core/thread.c]: @pfl  ? This is defined only kernel/include/kernel_structs.h. I agree with @andrewboie, this should have zephry's namespace.

Comment (pfl) [reply arch/arc/core/thread.c]: _new_thread_init() is only defined in ./kernel/include/kernel_structs.h and there it is defined 'static ALWAYS_INLINE'. According to the initial understanding, static functions shall only have their initial underscores removed.
You are now saying that even though this (and other, there are more similar cases) function is static, it is anyway treated as a "global" and shall therefore start with 'z_' ?
Is this special handling only applicable to ./kernel/include/kernel_structs.h or what is thinking here?

File: scripts/gen_syscall_header.py (hunk 4 above)
Comment (pfl) [reply scripts/gen_syscall_header.py]: Updated.

File: arch/arc/core/prep_c.c (hunk 5 above)
Comment (pfl) [reply arch/arc/core/prep_c.c]: Yes, fixed.

**Review by pfalcon — APPROVED** [review]
Ok, this is a change in the right direction apparently, so approving to get off my list. Leaving specific matters to other folks.

**Comment by pfl** [comment]
With the initial help of the scripts in #12086 and some manual updating, this patch survives 'sanitycheck -t net' when applied on master at commit 78b644ea3a07ec6393355e01ada929066c6d50ee.

All exported fuctions should now start with the prefix 'z_', even when they are marked static.

**Comment by pfl** [comment]
Updated to latest master.

File: arch/arc/core/irq_manage.c [hunk 6]
```diff
@@ -106,7 +106,7 @@ void _irq_spurious(void *unused)
 }
 
 #ifdef CONFIG_DYNAMIC_INTERRUPTS
-int _arch_irq_connect_dynamic(unsigned int irq, unsigned int priority,
+int z_arch_irq_connect_dynamic(unsigned int irq, unsigned int priority,
 			      void (*routine)(void *parameter), void *parameter,
```
Comment (finikorg) [inline arch/arc/core/irq_manage.c]: is alignment now broken?

File: arch/arc/core/mpu/arc_core_mpu.c [hunk 7]
```diff
@@ -104,15 +104,15 @@ void configure_mpu_mem_domain(struct k_thread *thread)
...
 {
 	return arc_core_mpu_get_max_domain_partition_regions();
 }
 
 /*
  * Reset MPU region for a single memory partition
  */
-void _arch_mem_domain_partition_remove(struct k_mem_domain *domain,
+void z_arch_mem_domain_partition_remove(struct k_mem_domain *domain,
 				       u32_t partition_id)
```
Comment (finikorg) [inline arch/arc/core/mpu/arc_core_mpu.c]: and here, probably other places

File: arch/nios2/core/thread.c [hunk 8]
```diff
@@ -31,17 +31,17 @@ struct init_stack_frame {
 };
 
 
-void _new_thread(struct k_thread *thread, k_thread_stack_t *stack,
+void z_new_thread(struct k_thread *thread, k_thread_stack_t *stack,
 		 size_t stack_size, k_thread_entry_t thread_func,
```
Comment (finikorg) [inline arch/nios2/core/thread.c]: I think in all cases where we change `_` to `z_` alignment is broken.

**Review by finikorg — DISMISSED** [review]
I think we need to correct indentation.

**Comment by pfl** [comment]
Indentation fixed with uncrustify. Fixed a lot more than just the renamed functions.

**Review by tyrantlizard77 — APPROVED** [review]

**Comment by pfl** [comment]
Just wondering what shippable is trying to tell me:
"geninfo: ERROR: cannot read sanity-out/native_posix/!"

File: arch/arc/core/mpu/arc_core_mpu.c (hunk 7 above)
Comment (pfl) [reply arch/arc/core/mpu/arc_core_mpu.c]: Done.

File: arch/arc/core/prep_c.c (hunk 1 above)
Comment (pfl) [reply arch/arc/core/prep_c.c]: Fixed.

File: arch/arc/core/thread.c (hunk 2 above)
Comment (pfl) [reply arch/arc/core/thread.c]: Fixed.

**Comment by pfl** [comment]
Hmm, `.section.text._handlermodeexit' by shippable is not touched by any patch. Is there something being generated or what should the conclusion be?

**Comment by aescolar** [comment]
> Just wondering what shippable is trying to tell me:
"geninfo: ERROR: cannot read sanity-out/native_posix/!"

That is in the coverage report generation.
Things have failed so miserably before, as to not have generated that folder at all.

File: arch/arc/include/kernel_arch_func.h (hunk 3 above)
Comment (pfl) [reply arch/arc/include/kernel_arch_func.h]: Naming updated.

**Review by ceolin — APPROVED** [review]

**Pull request merged**